src/Common/src/Interop/Unix/libssl/Interop.SafeSslHandle.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class libssl
+    {
+        internal sealed class SafeSslContextHandle : SafeHandle
+        {
+            public SafeSslContextHandle(IntPtr method)
+                : base(IntPtr.Zero, true)
+            {
+                handle = SSL_CTX_new(method);
+            }
+
+            public override bool IsInvalid
+            {
+                get { return handle == IntPtr.Zero; }
+            }
+
+            protected override bool ReleaseHandle()
+            {
+                SSL_CTX_free(handle);
+                return true;
+            }
+        }
+
+        internal sealed class SafeSslHandle : SafeHandle
+        {
+            private readonly SafeBioHandle _readBio;
+            private readonly SafeBioHandle _writeBio;
+            private bool _isServer;
+
+            public bool IsServer
+            {
+                get {return _isServer;}
+            }
+
+            public SafeBioHandle InputBio
+            {
+                get
+                {
+                    return _readBio;
+                }
+            }
+
+            public SafeBioHandle OutputBio
+            {
+                get
+                {
+                    return _writeBio;
+                }
+            }
+
+            public SafeSslHandle(SafeSslContextHandle context, bool isServer)
+                : base(IntPtr.Zero, true)
+            {
+                _isServer = isServer;
+
+                IntPtr memMethod = Interop.libcrypto.BIO_s_mem();
+                _readBio = Interop.libcrypto.BIO_new(memMethod);
+                _writeBio = Interop.libcrypto.BIO_new(memMethod);
+                
+                IntPtr tempHandle = SSL_new(context);
+                if (tempHandle == IntPtr.Zero)
+                {
+                    return;
+                }
+
+                // After SSL_set_bio, the BIO handles are owned by SSL pointer
+                // and are automatically freed by SSL_free. To prevent a double
+                // free, we need to keep the ref counts bumped up till SSL_free
+                bool gotReadRef = false, gotWriteRef = false;
+                _readBio.DangerousAddRef(ref gotReadRef);
+                _writeBio.DangerousAddRef(ref gotWriteRef);
+                SSL_set_bio(tempHandle, _readBio.DangerousGetHandle(), _writeBio.DangerousGetHandle());
```

Comment:
> In a static method, the try/catch would not be needed since ThreadAbortEx is not possible, is that right?

I don't know of any reason why a static method would be immune to a ThreadAbortException.  The example on [MSDN](https://msdn.microsoft.com/en-us/library/system.threading.threadabortexception%28v=vs.110%29.aspx) even uses only static methods for its demonstration.  The only thing that I know of that's immune is a Constrained Execution Region, which we don't have in CoreFX/CoreCLR.

That said, it seems like we might not have ThreadAbortException in CoreFX/CoreCLR (@janvorli or @stephentoub am I missing something here?).

> if SSL_new returns a SafeSslHandle, ctor signature would change, correct?

Yes, the constructor would likely go to `private SafeSslHandle()`.

> I don't see a clear benefit.

Getting native resources tracked by a SafeHandle as early as possible is always good.  The window from the time where you get the IntPtr to the time that it's properly tracked by a SafeHandle is a window for leaking native resources. The more code that can go between them, the more chances of the leak.  By having the Marshaller do the SafeHandle wrapping during the P/Invoke return you're about as immune to leaking as you can get.